### PR TITLE
Test 3: TicTacToe

### PR DESCRIPTION
**Pull request by Ayisha-Ahamed**

### CODE REVIEW

File: Main.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+// Test.c
+// Program on branch Test3.1
+// Program that implements TicTacToe.c.
+// ------------------------------------------------------------------------------------------------
+#include <stdio.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <io.h>
+#include <fcntl.h>
+#include "TicTacToe.h"
```
Comment (sanjivtmm) [inline Main.c]: It is a program you are implementing not a library. So it does not need to split into separate header and code files.

File: Main.c [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+static wchar_t Get_Input (wchar_t* prompt) {
```
Comment (sanjivtmm) [inline Main.c]: No comment/description of the function.
Nested while loop is making the logic obscure.
Use of continue seems suspect.

File: Main.c [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+   if (isAutoPlay) {
+      int num = 4;
+      while (arr[num] != L' ') num = rand () % 9;
+      return num;
+   }
+   while (true) {
+      wprintf (L"Enter box number (1-9): ");
+      wchar_t c = getwchar ();
+      while (getwchar () != L'\n');
+      if (c < L'1' || c > L'9' || arr[c - L'1'] != ' ') continue;
```
Comment (sanjivtmm) [inline Main.c]: Avoid use of continue. Better to reverse the condition and exit.

File: Main.c [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      if (l == 9) break;
+      wprintf (L"\n\u2523\u2501\u2501\u2501\u254B\u2501\u2501\u2501\u254B\u2501\u2501\u2501\u252B\n\u2503");
+   }
+   wprintf (L"\n\u2517\u2501\u2501\u2501\u253B\u2501\u2501\u2501\u253B\u2501\u2501\u2501\u251B\n");
+}
+
+static int Get_Box_Num (wchar_t* arr, bool isAutoPlay) {
+   if (isAutoPlay) {
+      int num = 4;
+      while (arr[num] != L' ') num = rand () % 9;
```
Comment (sanjivtmm) [inline Main.c]: Nice idea to use random as the first level of auto play!

File: TicTacToe.c [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#include <stdio.h>
+#include <stdbool.h>
+#include "TicTacToe.h"
+
+#define DIA1 6
+#define DIA2 7
+
+int TicTacToe (wchar_t* grid, int index, bool isPlayer1) {
+   // 3 x 3 Grid consists of three rows, three columns and two diagonals(8 possibilities)
+   static int Count[8][2] = {
```
Comment (sanjivtmm) [inline TicTacToe.c]: Interesting data structure.

**Comment by Dakshita-Minocha** [comment]
Please assign PRs to yourself before committing and pushing.

File: TicTacToe.c [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+// Gets two non-space unique symbols to represent player 1 and 2.
+static void Get_Symbol (char* s1, char* s2) {
+   *s1 = Get_Input (L"Enter symbol 1:");
+   *s2 = Get_Input (L"Enter symbol 2:");
+   while (*s1 == *s2) *s2 = Get_Input (L"Please enter a unique symbol. Enter symbol 2:");
+}
+
+// Prints 3 x 3 game board.
+static void PrintBoard (wchar_t* grid) {
+   wprintf (L"\u250F\u2501\u2501\u2501\u2533\u2501\u2501\u2501\u2533\u2501\u2501\u2501\u2513\n\u2503");
```
Comment (Dakshita-Minocha) [inline TicTacToe.c]: You can directly copy paste symbols. It is more readable than corresponding unicode.

File: TicTacToe.c [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+   for (int i = 0, l = 0; i < 3; i++) {
+      for (int k = 0; k < 3; l++, k++) wprintf (L" %wc \u2503", grid[l]);
+      if (l >= 9) break;
```
Comment (Dakshita-Minocha) [inline TicTacToe.c]:
```suggestion
   for (int i = 0, l = 0; i < 3 || l >= 9; i++) {
      for (int k = 0; k < 3; l++, k++) wprintf (L" %wc \u2503", grid[l]);
 

```

File: TicTacToe.c [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+int main () {
+   int val = _setmode (_fileno (stdout), _O_U8TEXT);
+   wchar_t grid[9] = { L' ',L' ',L' ' ,L' ' ,L' ' ,L' ' ,L' ' ,L' ',L' ' };
+   char P1, P2;
+   Get_Symbol (&P1, &P2);
+   bool isPlayer1 = true, isAutoPlay = false;
+   wprintf (L"Enter 1 to play against the computer. Enter 2 for dual player mode.\n");
+   if (ModeChoice ()) isAutoPlay = true;
```
Comment (Dakshita-Minocha) [inline TicTacToe.c]:
```suggestion
    isAutoPlay = ModeChoice ();

```

File: TicTacToe.c [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+   for (int i = 0; i < 9; i++, isPlayer1 = !isPlayer1) {
+      wprintf (L"%ls :\n", isPlayer1 ? L"Player 1" : L"Player 2");
```
Comment (Dakshita-Minocha) [inline TicTacToe.c]:
```suggestion
      wprintf (L"Player %d :\n", isPlayer1 );

```
isPlayer1 can be an int with 1 and 2 as possible values.
Check if this simplifies implementation.

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]

File: TicTacToe.c [hunk 10]
```diff
@@ -18,7 +18,7 @@
 #define DIA1 6
 #define DIA2 7
 
-static int TicTacToe (wchar_t* grid, int index, bool isPlayer1) {
+static int TicTacToe (char* grid, int index, int Player) {
```
Comment (Dakshita-Minocha) [inline TicTacToe.c]:
```suggestion
static int TicTacToe (char* grid, int index, int player) {

```

File: TicTacToe.c [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+#include <io.h>
+#include <fcntl.h>
+#include <string.h>
+
+#define DIA1 6
+#define DIA2 7
+
+static int TicTacToe (char* grid, int index, int Player) {
+   // 3 x 3 Grid consists of three rows, three columns and two diagonals(8 possibilities)
+   static int Count[8][2] = {
```
Comment (Dakshita-Minocha) [inline TicTacToe.c]: static variables start with s.
```suggestion
   static int sCount[8][2] = {

```

File: TicTacToe.c [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+   wprintf (L"┏━━━┳━━━┳━━━┓\n┃");
+   for (int row = 0, index = 0; row < 3; row++) {
+      for (int col = 0; col < 3 && index < 9; index++, col++) wprintf (L" %wc ┃", grid[index]);
+      if (row != 2) wprintf (L"\n┣━━━╋━━━╋━━━┫\n┃");
+   }
+   wprintf (L"\n┗━━━┻━━━┻━━━┛\n");
+}
+
+// Returns the index of the box number within the range of 0-8.
+static int Get_Box_Num (char* arr, bool isAutoPlay) {
```
Comment (Dakshita-Minocha) [inline TicTacToe.c]: Please follow naming conventions and read through a program before you submit.

File: TicTacToe.c [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+   if (isAutoPlay) {
+      int num = 4;
+      while (arr[num] != L' ') num = rand () % 9;
+      return num;
+   }
+   while (true) {
+      char boxNum = Get_Input (L"Enter box number (1-9):");
+      if (boxNum < L'1' || boxNum > L'9' || arr[boxNum - L'1'] != L' ') {
+         wprintf (L"Entered box number is either occupied or invalid\nExiting the program.....");
+         return -1;
```
Comment (Dakshita-Minocha) [inline TicTacToe.c:92]: Why return here? Why not handle the scenario?

**Review by Dakshita-Minocha — APPROVED** [review]

**Pull request closed without merging**